pep-9999.rst

```diff
@@ -0,0 +1,937 @@
+``foo`` to another thread will make the assignment ``x = 1`` a potential
+data race.
+
+We consider freezing of a function to freeze that function's
+meaning at that point in time. In the code above, that means that
+``foo`` gets its own copy of ``x`` which will have value of the enclosing
+``x`` at the time of freezing, in this case 0.
+
+Thus, the assignment ``x = 1`` is still permitted as it will not affect
+``foo``, and it may therefore not contribute to a data race. Furthermore,
+the result of calling ``foo()`` will be 0 -- not 1!
+
+This is implemented by having ``x`` in ``foo`` point to a fresh
+cell and then freezing the cell (and similar for global capture).
+Note that this also prevents ``x`` from being reassigned.
+
+We believe that this design is a sweet-spot that is intuitive and
+permissive. Note that we will treat freezing functions that
+capture enclosing state in the same way regardless of whether the
+enclosing state is another function or the top-level (i.e., the
+enclosing scope is ``globals()``).
+
+(A **rejected alternative** is to freeze ``x`` in the
+enclosing scope. This is problematic when a captured variable is
+in ``globals()`` and also rejects more programs.)
+
+Now consider freezing the following function:
+
+.. code-block:: python
+
+  from immutable import freeze
+
+  def example2():
+      x = 0
+      def foo(a = False):
+          if a:
+              a = a + 1 # Note: updating local variables work, even in a frozen function
+              return a
+          else:
+              x = x + 1
+              return x
+
+      freeze(foo)
+      foo(41)  # OK, returns 42
+      foo()    # Throws NotWriteableError
+
+  example2()
+
+This example illustrates two things. The first call to ``foo(41)``
+shows that local variables on the frame of a frozen function are
+mutable. The second call shows that captured variables are not.
+Note that the default value of ``a`` will be frozen when ``foo``
+is frozen. Thus, the problem of side-effects on default values
+on parameters is avoided.
+
+Frozen function objects that access globals, e.g. through an
+explicit call to ``globals()``, will throw an exception when
+called.
+
+
+Implementation Details
+======================
+
+1. Add the ``immutable`` module, the ``NotWriteableError`` type, and
+   the ``NotFreezable`` type.
+2. Add the ``freeze(obj)`` function to the ``immutable`` module and
+   ensure that it traverses object references safely, including
+   cycle detection, and marks objects appropriately, and backs
+   out on failure, without partially freezing object graphs.
+3. Add the ``register_freezable(type)`` whitelisting function.
+4. Add the ``isfreezable(obj)`` function which checks that all
+   objects reachable from ``obj`` can be frozen.
+5. Add the ``register_freezable(type)`` function that is used to
+   whitelist types implemented as C extensions, permitting their
+   instances to be frozen.
```

Comment:
Are 3. and 5. supposed to have the same function name? Should we remove 3.

Reply:
```suggestion
3. Add the ``isfreezable(obj)`` function which checks that all
   objects reachable from ``obj`` can be frozen.
4. Add the ``register_freezable(type)`` function that is used to
   whitelist types implemented as C extensions, permitting their
   instances to be frozen.
```